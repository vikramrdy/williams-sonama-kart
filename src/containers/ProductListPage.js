import {
  CircularProgress,
  GridList,
  GridListTile,
  Paper,
  makeStyles,
  useMediaQuery,
  useTheme,
} from "@material-ui/core";
import React, { useEffect, useState } from "react";
import axios from "axios";

import CarouselModal from "../components/CarouselModal";
import Product from "../components/Product";
import { ProductListMockGroups } from "../__mockData__/ProductListMockGroups";

const useStyles = makeStyles(() => ({
  root: {
    background: "whitesmoke",
  },
  gridListTile: {
    border: "1px solid lightGray",
    borderRadius: 7,
  },
  noData: {
    textAlign: "center",
    paddingTop: 20,
  },
  progress: {
    margin: 16,
    position: "absolute",
    top: "50%",
    left: "50%",
  },
}));

const getFormattedProduct = (p) => {
  return {
    id: p.id,
    name: p.name.replace(/&amp;/g, "&"),
    priceRange: p.priceRange,
    lowPrice: p.priceRange?.selling.low,
    highPrice: p.priceRange?.selling.high,
    hero: p.hero,
    thumbnail: p.images,
  };
};

const getFormattedProductGroups = (groups) => {
  return groups.map((p) => getFormattedProduct(p));
};

const ProductListPage = () => {
  const classes = useStyles();
  const theme = useTheme();
  const isMobileResolution = useMediaQuery(theme.breakpoints.down("xs"));

  const [thumbnail, setThumbnail] = useState(null);
  const [productList, setProductList] = useState(null);
  const [loading, setLoading] = useState(true);

  const getProductList = () => {
    axios
      .get(
        `https://www.westelm.com/services/catalog/v4/category/shop/new/all-new/index.json`
      )
      .then((response) => {
        const productGroupList = getFormattedProductGroups(
          response?.data?.groups
        );
        setProductList(productGroupList);
        setLoading(false);
      })
      .catch(() => {
        // To load product list from mock when fetching data through service fails
        const productGroupList = getFormattedProductGroups(
          ProductListMockGroups.groups
        );
        setProductList(productGroupList);
        setLoading(false);
      });
  };

  useEffect(() => {
    getProductList();
  }, []);

  if (loading)
    return <CircularProgress className={classes.progress} size={50} />;

  if (!productList) return <div className={classes.noData}>No Data Found</div>;

  return (
    <Paper classes={{ root: classes.root }} elevation={0}>
      <GridList
        cols={isMobileResolution ? 1 : 3}
        cellHeight={300}
        spacing={32}
        style={{ margin: 20 }}
      >
        {productList.map((product) => (
          <GridListTile
            key={product.id}
            classes={{ tile: classes.gridListTile }}
            onClick={() => setThumbnail(product.thumbnail)}
          >
            <Product data={product} />
          </GridListTile>
        ))}
      </GridList>
      {Boolean(thumbnail) && (
        <CarouselModal
          thumbnail={thumbnail}
          open={Boolean(thumbnail)}
          onClose={() => setThumbnail(null)}
        />
      )}
    </Paper>
  );
};

export default ProductListPage;
