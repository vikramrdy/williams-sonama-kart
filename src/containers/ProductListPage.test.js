import { render } from "@testing-library/react";
import React from "react";

import ProductListPage from "./ProductListPage";

describe("ProductListPage", () => {
  it("should render the product list component", () => {
    const { asFragment } = render(<ProductListPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
