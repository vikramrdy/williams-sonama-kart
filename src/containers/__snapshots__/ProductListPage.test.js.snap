// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListPage should render the product list component 1`] = `
<DocumentFragment>
  <div
    class="MuiCircularProgress-root makeStyles-progress-4 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 50px; height: 50px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;
